source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,410 @@
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_map.h"
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::LowerCaseString;
+using ::Envoy::Http::Protocol;
+using HeaderValidatorFunction =
+    HeaderValidator::HeaderValueValidationResult (Http2HeaderValidator::*)(const HeaderString&);
+
+struct Http2ResponseCodeDetailValues {
+  const std::string InvalidTE = "uhv.http2.invalid_te";
+  const std::string ConnectionHeaderSanitization = "uhv.http2.connection_header_rejected";
+};
+
+using Http2ResponseCodeDetail = ConstSingleton<Http2ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/2 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 9110 <https://www.rfc-editor.org/rfc/rfc9110.html> HTTP Semantics
+ * RFC 9112 <https://www.rfc-editor.org/rfc/rfc9112.html> HTTP/1.1
+ * RFC 9113 <https://www.rfc-editor.org/rfc/rfc9113.html> HTTP/2
+ *
+ */
+Http2HeaderValidator::Http2HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http2HeaderValidator::validateMethodHeader},
+      {":authority", &Http2HeaderValidator::validateAuthorityHeader},
+      {":scheme", &Http2HeaderValidator::validateSchemeHeader},
+      {":path", &Http2HeaderValidator::validatePathHeaderCharacters},
+      {"te", &Http2HeaderValidator::validateTEHeader},
+      {"content-length", &Http2HeaderValidator::validateContentLengthHeader},
+  };
+
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  auto validator_it = kHeaderValidatorMap.find(key_string_view);
+  if (validator_it != kHeaderValidatorMap.end()) {
+    const auto& validator = validator_it->second;
+    return (*this.*validator)(value);
+  }
+
+  if (key_string_view.at(0) != ':') {
+    // Validate the (non-pseudo) header name
+    auto name_result = validateGenericHeaderName(key);
```

Comment:
do we need to make sure that the host request header isn't set for HTTP/2 and above, potentially causing problems for downgrade to HTTP/1.1?
I thought it was legal for responses but not for requests but I may be misremembering.

Reply:
I don't think that is an issue because of how nghttp2 and Envoy handle the `host` to `:authority` aliasing. I've tried a couple scenarios and I cannot see how a request header map with a `host` header would reach UHV.

- Specifying both `host` and `:authority`: The `host` value is overrides `:authority` and is removed
- Specifying only `host`: The `host` value is moved to `:authority`